transform/minifyhtml.go

```diff
@@ -0,0 +1,62 @@
+// Copyright 2016 The Hugo Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package transform
+
+import (
+	"bytes"
+	"io"
+	"regexp"
+
+	"github.com/tdewolff/minify"
+	"github.com/tdewolff/minify/css"
+	"github.com/tdewolff/minify/html"
+	"github.com/tdewolff/minify/js"
+	"github.com/tdewolff/minify/json"
+)
+
+var minifier *minify.M
+
+func init() {
+	m := minify.New()
+	m.AddFunc("text/css", func(m *minify.M, w io.Writer, r io.Reader, params map[string]string) error {
+		var buf bytes.Buffer
+		if _, err := buf.ReadFrom(r); err != nil {
+			return err
+		}
+
+		// css.Minify will minify "0s" to be "0" thus breaking the AMP boiler plate.
+		if bytes.Contains(buf.Bytes(), []byte("-amp-start")) {
+			_, err := buf.WriteTo(w)
+			return err
+		}
+
+		return css.Minify(m, w, &buf, params)
+	})
+	m.Add("text/html", &html.Minifier{
+		KeepDefaultAttrVals: true,
+		KeepWhitespace:      true,
+	})
+	m.AddFunc("text/javascript", js.Minify)
+	m.AddFuncRegexp(regexp.MustCompile("[/+]json$"), json.Minify)
+	minifier = m
+}
+
+// MinifyHTML minifies the HTML and the embedded CSS, JS and JSON.
+func MinifyHTML(ct contentTransformer) {
+	var buf bytes.Buffer
+	buf.Write(ct.Content())
```

Comment:
I will look closer at this PR later, but 3 comments for now:

* If you need a `io.Reader` from a `byte` slice, use `bytes.NewReader`, it is 200 times faster and allocates no memory.
* Not all content produced by Hugo is HTML, a user can set the `url` in to `index.json` as an example.
* You need to pull in the dependencies into the vendor file for the CI build to go green. See the `govendor` doc.